File: nestty/assurance.ts
/**
 * Assurance validation logic for NesTTY.
 *
 * Assurance is a Tier 1 role that validates completed agent work in two layers:
 * 1. Verify the agent's self-verification (Ralph Wiggum Loop) actually ran
 * 2. Independently score each @success_criteria item
 *
 * Gate: 95% weighted score. Below → return to agent with gap analysis.
 * Above → forward to QA/Synthesizer assembly queue.
 *
 * This module provides the prompt templates and scoring logic.
 * The actual LLM evaluation happens via the Assurance agent's PTY session.
 */

/** Validation result for a single criterion */
export interface CriterionResult {
  criterion: string;
  passed: boolean;
  score: number;       // 0-100
  feedback?: string;
}

/** Full validation verdict */
export interface ValidationVerdict {
  taskId: string;
  passed: boolean;
  overallScore: number;  // 0-100
  criteriaResults: CriterionResult[];
  selfVerificationChecked: boolean;
  selfVerificationValid: boolean;
  gaps?: string;
  feedback?: string;
  timestamp: string;
}

const PASS_THRESHOLD = 95;

/**
 * Build the validation prompt for the Assurance agent.
 * This gets injected into the Assurance agent's PTY when a task
 * is submitted for validation.
 */
export function buildValidationPrompt(task: {
  id: string;
  title?: string;
  description: string;
  successCriteria: string[];
  result: string;
  selfVerification?: string;
}): string {
  const criteriaList = task.successCriteria
    .map((c, i) => `  ${i + 1}. ${c}`)
    .join('\n');

  // Identity, scoring rubric, and validation process are in the Go system prompt
  // (act-agent/internal/llm/prompt/assurance.go). This turn prompt provides the DATA.
  return [
    `VALIDATION REQUEST — Task: ${task.title || task.id}`,
    ``,
    `## Task Description`,
    task.description,
    ``,
    `## Success Criteria`,
    criteriaList,
    ``,
    `## Agent's Output (Result)`,
    task.result.substring(0, 4000), // cap to avoid prompt explosion
    ``,
    task.selfVerification ? [
      `## Agent's Self-Verification`,
      task.selfVerification.substring(0, 2000),
      ``,
    ].join('\n') : '',
    `Validate this task now. Respond with the JSON verdict format.`,
  ].join('\n');
}

/**
 * Parse the Assurance agent's JSON response into a ValidationVerdict.
 * Returns null if parsing fails.
 */
export function parseValidationResponse(
  taskId: string,
  rawResponse: string
): ValidationVerdict | null {
  try {
    // Try to extract JSON from the response (agent might wrap it in text)
    const jsonMatch = rawResponse.match(/\{[\s\S]*"criteriaResults"[\s\S]*\}/);
    if (!jsonMatch) return null;

    const parsed = JSON.parse(jsonMatch[0]);

    const criteriaResults: CriterionResult[] = (parsed.criteriaResults || []).map(
      (cr: any) => ({
        criterion: String(cr.criterion || ''),
        passed: Boolean(cr.passed),
        score: Number(cr.score) || 0,
        feedback: cr.feedback ? String(cr.feedback) : undefined,
      })
    );

    const overallScore = Number(parsed.overallScore) || 0;

    return {
      taskId,
      passed: overallScore >= PASS_THRESHOLD,
      overallScore,
      criteriaResults,
      selfVerificationChecked: true,
      selfVerificationValid: Boolean(parsed.selfVerificationValid),
      gaps: parsed.gaps ? String(parsed.gaps) : undefined,
      feedback: parsed.feedback ? String(parsed.feedback) : undefined,
      timestamp: new Date().toISOString(),
    };
  } catch {
    return null;
  }
}

/**
 * Build a gap analysis prompt for returning failed work to an agent.
 * This tells the agent what specifically needs fixing.
 */
export function buildGapAnalysisPrompt(verdict: ValidationVerdict): string {
  const failedCriteria = verdict.criteriaResults
    .filter(cr => !cr.passed)
    .map(cr => `- ${cr.criterion} (score: ${cr.score}/100): ${cr.feedback || 'no details'}`)
    .join('\n');

  return [
    `VALIDATION FAILED — Score: ${verdict.overallScore}/100 (need ${PASS_THRESHOLD}+)`,
    ``,
    `## Failed Criteria`,
    failedCriteria || '(none individually failed, but overall score too low)',
    ``,
    verdict.gaps ? `## Gap Analysis\n${verdict.gaps}\n` : '',
    `Fix the specific gaps above, then re-submit with \`act task submit-for-validation\`.`,
  ].join('\n');
}

File: act-agent/internal/llm/prompt/assurance.go
package prompt

import (
	"fmt"

	"github.com/paradiselabs-ai/ACT/act-agent/internal/llm/models"
)

// AssurancePrompt returns the system prompt for the Assurance role.
func AssurancePrompt(provider models.ModelProvider) string {
	cli := actCLICommands("assurance")
	if provider == models.ProviderACP {
		cli = actCLICommandsACP("assurance")
	}
	envInfo := getEnvironmentInfo()
	return fmt.Sprintf("%s\n\n%s\n\n%s\n\n%s\n\n%s", baseAssurancePrompt, cli, communicationProtocol(), coordinationConstraints("assurance"), envInfo)
}

const baseAssurancePrompt = `You are Assurance — you validate that completed agent work meets its success criteria.

# Identity & Role
You operate in the NesTTY conversation window (Tier 1 — interactive). Your job:
1. Receive tasks submitted for validation from the orchestrator
2. Perform two-layer validation on each task
3. Return a structured verdict (PASS at 95%+ or FAIL with gap analysis)

# Two-Layer Validation

## Layer 1: Self-Verification Check (Ralph Wiggum Loop)
Did the agent claim to verify their own work? This is a quality signal only — an agent that didn't self-verify is more likely to have gaps. Never accept the agent's claims as evidence on their own.

## Layer 2: Independent Tool-Based Criteria Scoring
Treat the agent's submitted result as an UNVERIFIED CLAIM. Use your ` + "`view`" + ` and ` + "`grep`" + ` tools to confirm or refute each @success_criteria item against the actual project working directory. A criterion scored without a tool-verified reasoning line MUST fail.

The PASS threshold is 95% (weighted average of all criteria scores).

# Validation Process
Each validation request includes the project working directory, the task description with @success_criteria, and the agent's submitted claim. Open files with ` + "`view`" + `, scan with ` + "`grep`" + `, decide per criterion.

Respond with EXACTLY this JSON shape (no markdown fences, no surrounding prose):
{
  "selfVerificationValid": true/false,
  "criteriaResults": [
    {"criterion": "criterion text", "passed": true/false, "reasoning": "ran `+"`"+`view path/to/file`+"`"+` → saw <exact evidence>"}
  ],
  "score": 0-100,
  "passed": true/false,
  "gaps": "specific gap analysis if failed, null if passed",
  "feedback": "overall assessment"
}

# When Validation Fails (< 95%)
Your gap analysis is sent back to the agent automatically. Be specific:
- Which criteria failed and why
- What exactly needs to change
- Do NOT rewrite the solution — describe what's wrong

The agent will rework and re-submit. You may see the same task multiple times.

# act_cli — your ONLY shell-style tool
Allowed subcommands: validation, log, status. Any other subcommand is rejected. Use the ` + "`view`" + ` and ` + "`grep`" + ` tools to read submitted source files directly.
- ` + "`{\"subcommand\":\"validation\",\"args\":[\"queue\"]}`" + ` — pending-validation list
- ` + "`{\"subcommand\":\"log\",\"args\":[\"--tail\",\"40\"]}`" + ` — context on what happened before submission
- ` + "`{\"subcommand\":\"status\"}`" + ` — system snapshot

# Output Style
Be precise and objective. Score based on evidence, not feeling. If a criterion is ambiguous,
interpret it reasonably and note the ambiguity in your feedback. When in doubt, be strict —
it's better to send work back for improvement than to pass substandard output.

# Refusal Clause
If the user message is not a validation request (no @success_criteria block, no submitted agent result, no working directory), respond with the empty string. Do NOT comment on system state, agent workload, task assignment, or decisions. Validation is your ONLY output. Anything else is the Planner's job.
If a validation request arrives with ZERO success criteria, never pass it: the verdict MUST be {"passed": false, "score": 0, "gaps": "missing success criteria", "criteriaResults": []}. No criteria means nothing to verify against — that is a validation failure, not a free pass.`
